### PR TITLE
DT-23494 DT-23495 Strengthen test assertions and correct internal tree node state

### PR DESCRIPTION
**Pull request by travissanderson-wf**
We (actually Daniel) identified a number of places where nodes were assigned the wrong parent, where nodes were left empty in the tree, and where nodes had the incorrect bounds. This PR fixes the bugs and adds assert utils for asserting valid tree state.

A happy side effect of having accurate tree state is that the tree is more performant, particularly for `remove`. We've also included a few micro-optimizations to eek out more performance.

  | master | https://github.com/Workiva/r_tree/pull/67 | vs master
-- | -- | -- | --
Insert 100 | 13,947.10 | 12,089.22 | 13.32%
Insert 1000 | 373,583.33 | 234,561.11 | 37.21%
Insert 10000 | 7,818,975.00 | 4,998,650.00 | 36.07%
Load 100 | 2,423.35 | 1,741.87 | 28.12%
Load 1000 | 42,995.06 | 35,771.34 | 16.80%
Load 10000 | 538,250.00 | 466,090.00 | 13.41%
Remove 5k | 403,643.34 | 142,043.33 | 64.81%
Search/Iterate (using Insert) 100 | 22,713.39 | 21,403.30 | 5.77%
Search/Iterate (using Insert) 1000 | 85,600.00 | 79,234.62 | 7.44%
Search/Iterate (using Insert) 10000 | 245,411.06 | 219,760.00 | 10.45%
Search/Iterate (using Load) 100 | 28,896.38 | 29,758.09 | -2.98%
Search/Iterate (using Load) 1000 | 102,107.48 | 103,170.00 | -1.04%
Search/Iterate (using Load) 10000 | 260,831.19 | 252,800.00 | 3.08%

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/r_tree/rect_util.dart
Comment (danielway-wk) [inline lib/src/r_tree/rect_util.dart:1]: If we can't use these utilities for both test/prod code, then I think we should just move `getMinimumBoundingRectangle` to `r_tree_test.dart`. I moved away from it because I was concerned the use of closures was negatively impacting performance.

File: lib/src/r_tree/non_leaf_node.dart [hunk 1]
```diff
@@ -119,21 +116,13 @@ class NonLeafNode<E> extends Node<E> {
...
 
-  _convertToLeafNode() {
-    var nonLeafParent = parent as NonLeafNode<E>?;
-    if (nonLeafParent == null) return;
-
-    var newLeafNode = LeafNode<E>(this.branchFactor);
-    newLeafNode.include(this);
-    nonLeafParent.removeChild(this);
-    nonLeafParent.addChild(newLeafNode);
-  }
```
Comment (danielway-wk) [inline lib/src/r_tree/non_leaf_node.dart:130]: For this method, see commentary on 4eca138423e030511d3ab1afd12f114da3a9510f.

File: test/r_tree/r_tree_test.dart [hunk 2]
```diff
@@ -213,22 +214,26 @@ main() {
...
         data.forEach((RTreeDatum item) {
           tree.remove(item);
         });
-        assertTreeHeightValidity(tree);
+        assertTreeValidity(tree);
 
         items = tree.search(Rectangle(0, 0, 50, 50));
         expect(items.length, equals(0));
 
+        expect(tree.currentRootNode, isA<LeafNode<dynamic>>());
```
Comment (danielway-wk) [inline test/r_tree/r_tree_test.dart:232]: This asserts that `RTree._resetRoot` resets a non-leaf to being a leaf in the one edge case.

File: benchmark/web_benchmarks.dart [hunk 3]
```diff
@@ -3,7 +3,16 @@ import 'dart:html';
...
 main() {
-  querySelector('#runButton')!.onClick.listen((_) {
+  final button = querySelector('#runButton')! as ButtonElement;
+  button.onClick.listen((_) async {
+    button.disabled = true;
+    final text = button.innerText;
+    button.innerText = 'Running';
+    await window.animationFrame;
+    await window.animationFrame;
+    await window.animationFrame;
```
[the remark this answers is not in the document]
Comment (travissanderson-wf) [reply benchmark/web_benchmarks.dart:13]: I guess so! I thought just one ought to be enough but it wasn't, so I went straight to 3.

File: lib/src/r_tree/node.dart [hunk 4]
```diff
@@ -94,23 +87,21 @@ abstract class Node<E> implements RTreeContributor {
...
-      _minimumBoundingRect = null;
-      return noMBR;
+      return _minimumBoundingRect = noMBR;
     }
 
-    var newBoundingRect = children.first.rect;
-    for (var child in children.skip(1)) {
-      newBoundingRect = newBoundingRect.boundingBox(child.rect);
+    var updatedBoundingRect = children[0].rect;
+    for (var i = 1; i < children.length; i++) {
```
[the remark this answers is not in the document]
Comment (travissanderson-wf) [reply lib/src/r_tree/node.dart:100]: unfortunately, yes

File: lib/src/r_tree/non_leaf_node.dart (hunk 1 above)
Comment (travissanderson-wf) [reply lib/src/r_tree/non_leaf_node.dart:130]: thanks, I intended to migrate the comments but hadn't yet :D

File: lib/src/r_tree/node.dart [hunk 5]
```diff
@@ -94,23 +87,21 @@ abstract class Node<E> implements RTreeContributor {
...
-    _minimumBoundingRect = _minimumBoundingRect == null ? item.rect : rect.boundingBox(item.rect);
+    _minimumBoundingRect = _minimumBoundingRect == noMBR ? item.rect : rect.boundingBox(item.rect);
   }
 
   /// Recalculated the bounding rectangle of this node
   Rectangle updateBoundingRect() {
     if (children.isEmpty) {
-      _minimumBoundingRect = null;
-      return noMBR;
+      return _minimumBoundingRect = noMBR;
```
Comment (jacobzuo-wk) [inline lib/src/r_tree/node.dart:96]: I think the old way of updating the value then return the value is clearer, merging the two operations into a single line makes it read too similiar to returning a boolean value here, i.e., `_minimumBoundingRect == noMBR`.

File: lib/src/r_tree/node.dart [hunk 6]
```diff
@@ -94,23 +87,21 @@ abstract class Node<E> implements RTreeContributor {
...
-    for (var child in children.skip(1)) {
-      newBoundingRect = newBoundingRect.boundingBox(child.rect);
+    var updatedBoundingRect = children[0].rect;
+    for (var i = 1; i < children.length; i++) {
+      updatedBoundingRect = updatedBoundingRect.boundingBox(children[i].rect);
     }
 
-    _minimumBoundingRect = newBoundingRect;
-    return _minimumBoundingRect!;
+    return _minimumBoundingRect = updatedBoundingRect;
```
Comment (jacobzuo-wk) [inline lib/src/r_tree/node.dart:104]: Same comment as above, let's split this logic into two separate lines.

File: lib/src/r_tree/node.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (jacobzuo-wk) [reply lib/src/r_tree/node.dart:100]: #nit: I think we can avoid worrying about skipping the first element by defining initializing `updatedBoundingRect` with a `null`.

```
    var updatedBoundingRect = children[0].rect;
    for (var i = 1; i < children.length; i++) {
        updatedBoundingRect = updatedBoundingRect.boundingBox(children[i].rect);
    }
}
```suggestion
    var updatedBoundingRect = null;
    for (var i = 1; i < children.length; i++) {
        updatedBoundingRect = updatedBoundingRect?.boundingBox(children[i].rect) ?? children[i].rect;
    }
```

File: lib/src/r_tree/node.dart (hunk 6 above)
Comment (travissanderson-wf) [reply lib/src/r_tree/node.dart]: I think that's a good call, done

File: lib/src/r_tree/node.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (travissanderson-wf) [reply lib/src/r_tree/node.dart:101]: I think I prefer the current code to the null-aware version, but that's just my subjective preference.

[the remark this answers is not in the document]
Comment (travissanderson-wf) [reply lib/src/r_tree/node.dart:101]: (sorry for the late reply, I thought you deleted your comment but it was hidden in a resolved thread, so I unresolved it)

[the remark this answers is not in the document]
Comment (jacobzuo-wk) [reply lib/src/r_tree/node.dart:101]: No worries, I don't have a strong opinion on this, I can go either direction.

**Review by seanstrickland-wf — APPROVED** [review]

**Review by jacobzuo-wk — APPROVED** [review]

**Comment by kentiverson-wf** [comment]
Starting QA...

**Comment by kentiverson-wf** [comment]
QA +1
- [x] Smoke tested a DPC build consuming these changes, No issues seen inserting / deleting table rows and applying cell formats.
- [x] Consumed these changes in DPC and ran through func test suite, all tests passed https://wf-skynet.appspot.com/apps/test/smithy/5165608/1
- [x] CI Passing

**Comment by travissanderson-wf** [comment]
@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM
